### ISSUE TITLE
`warning: unnecessary trailing semicolon` false positive in macro_rules

### ISSUE THREAD

**Issue by eNV25** [rust-lang/rust#117569]
I tried this code:

```rust
macro_rules! try_block {
    ($($stmt:stmt);+) => {
        (|| { $($stmt);+ })()
    };
}

pub struct Value;
pub struct Error;
type Result<T> = std::result::Result<T, Error>;

pub fn try_block() {
    assert!(try_block! {
        Result::Err(Error)?;
        Result::Ok(Value)
    }.is_err());
}
```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=5ffd8e513c639670f2523132fdf21f1c

I expected to see this happen: no warnings

Instead, this happened:

```
   Compiling playground v0.0.1 (/playground)
warning: unnecessary trailing semicolon
 --> src/lib.rs:3:23
  |
3 |         (|| { $($stmt);+ })()
  |                       ^ help: remove this semicolon
  |
  = note: `#[warn(redundant_semicolons)]` on by default

warning: `playground` (lib) generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 0.24s
```

There is in fact no extra semicolon, you can verify with the expand macros function the above playground link. Also with `cargo expand`.

### Meta

`rustc --version --verbose`:
```
rustc 1.73.0 (cc66ad468 2023-10-03)
binary: rustc
commit-hash: cc66ad468955717ab92600c770da8c1601a4ff33
commit-date: 2023-10-03
host: x86_64-unknown-linux-gnu
release: 1.73.0
LLVM version: 17.0.2
```

Also happens in playground, both stable and nightly.

</p>
</details>

**Comment by eggyal** [comment]
The macro expansion:

```rust
(|| { Result::Err(Error)?; ; Result::Ok(Value) })()
//                         ^ unnecessary semicolon
```

**Comment by eggyal** [comment]
The `stmt` metavariable includes the terminating semicolon.  Perhaps you want to omit the `;` from the repeating pattern, or use an `expr` metavariable instead?

**Comment by eggyal** [comment]
The `;` in the input results in the repetition being matched three times: the `stmt` metavariable respectively with `Result::Err(Error)?`, empty `` and `Result::Ok(Value)`.  You probably want to remove it from the input, or else use `expr` metavariables instead.
